// menashmq/dist/lib/connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connection = void 0;
const internal_1 = require("./internal");
class Connection {
    constructor(url, socketOptions) {
        this.url = url;
        this.socketOptions = socketOptions;
        this.connection = null;
    }
    async initialize() {
        if (this.connection) {
            await this.close();
        }
        this.connection = await internal_1.amqp.connect(this.url, this.socketOptions);
        this.connection.once('error', (err) => {
            this.connection = null;
            internal_1.default.reportError('connection', err);
        });
        this.connection.once('close', () => {
            this.connection = null;
            internal_1.default.reportError('connection', new Error('Connection closed'));
        });
    }
    async close() {
        if (!this.connection) {
            return;
        }
        const { connection } = this;
        this.connection = null;
        connection.removeAllListeners('error');
        connection.removeAllListeners('close');
        await connection.close().catch((err) => {
            console.log(`Warning: failed to close connection with error`, err);
        });
    }
    isConnected() {
        return !!this.connection;
    }
}
exports.Connection = Connection;
