// menashmq/dist/lib/internal.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pRetry = void 0;
const pRetry = require("p-retry");
exports.pRetry = pRetry;
exports.amqp = require("amqplib");
var assert_1 = require("assert");
Object.defineProperty(exports, "assert", { enumerable: true, get: function () { return assert_1.strict; } });
var events_1 = require("events");
Object.defineProperty(exports, "EventEmitter", { enumerable: true, get: function () { return events_1.EventEmitter; } });
Object.defineProperty(exports, "once", { enumerable: true, get: function () { return events_1.once; } });
var util_1 = require("util");
Object.defineProperty(exports, "isDeepStrictEqual", { enumerable: true, get: function () { return util_1.isDeepStrictEqual; } });
var utils_1 = require("../utils");
Object.defineProperty(exports, "trycatch", { enumerable: true, get: function () { return utils_1.trycatch; } });
Object.defineProperty(exports, "stringify", { enumerable: true, get: function () { return utils_1.stringify; } });
__exportStar(require("./connection"), exports);
__exportStar(require("./channel"), exports);
__exportStar(require("./queue"), exports);
__exportStar(require("./exchange"), exports);
__exportStar(require("./message"), exports);
__exportStar(require("./binding"), exports);
__exportStar(require("./client"), exports);
const client_1 = require("./client");
const client = new client_1.Client();
exports.default = client;
